doc/GUIDE.md

```diff
@@ -0,0 +1,186 @@
+
+libfsm provides two main tools: **`re`** takes patterns from command line, **`rx`** takes patterns from file.
+
+### 1. Validate the Regex
+
+Test behavior using any PCRE-compatible tool (e.g., [https://regex101.com/](https://regex101.com/)).
+
+### 2. Verify libfsm Compatibility
+
+```bash
+re -r pcre -l ast 'x*?'
+# Output: /x*?/:3: Unsupported operator
+
+rx -r pcre -l ast -d declined.txt 'x*?'
+# Unsupported character in declined.txt
+```
+
+If unsupported constructs exist, libfsm reports the failing location.
+
+### 3. Generate Code
+
+```bash
+re -p -r pcre -l rust -k str '^item-[A-Z]{3}$' > item_detector.rs
+```
+
+### 4. Multiple Patterns
+
+```bash
+# re - patterns from command line:
+re -p -r pcre -l go -k str '^x?a b+c$' '^x*def?$' '^x$'
+
+# rx - patterns from file:
+rx -p -r pcre -l vmc -k str -d skipped.txt patterns.txt > detectors.c
+```
+
+Both tools:
+* Combine all patterns into one function (like using `|` to join them)
+* Return `(bool, int)` - match status and pattern ID
+* Pattern ID is argument position for `re`, line number for `rx`
+* When encountering unsupported patterns: `rx` skips them to `-d` file and generates code with working patterns; `re` fails completely
+
+---
+
+### Flag Reference
+| Flag | Purpose                     | Common Options                             | Notes                                      |
+| ---- | --------------------------- | ------------------------------------------ | ------------------------------------------ |
+| `-r` | Select regex dialect        | `pcre`, `literal`, `glob`, `native`, `sql` | `pcre` supports the widest set of features |
+| `-l` | Choose output language      | `go`, `rust`, `vmc`, `llvm`, `wasm`, `dot` | Use `vmc` for `C` code, pipe `dot` into `idot` for visualization |
+| `-k` | Generated function I/O API  | `str`, `getc`, `pair`                      | `str` takes string, `pair` takes byte array, `getc` uses callback for streaming |
+| `-p` | Production mode             | *(no value)*                               | Generates optimized code                   |
+| `-d` | Output unsupported patterns | filename                                   | Only applies to `rx` (batch mode)          |
+
+For more detailed information on flags, see [include/fsm/options.h](../include/fsm/options.h) and the man pages (by running `build/man/re.1/re.1` after `bmake doc`).
+
+---
+
+## Writing Effective libfsm Patterns
+
+For additional regex best practices, see [Fastly's regex guide](https://www.fastly.com/documentation/reference/vcl/regex/#best-practices-and-common-mistakes).
+
+### 1. Replace Broad Wildcards
+
+Avoid `.*` whenever possible. Use negated character classes:
+
+| Avoid      | Better         |
+| ---------- | -------------- |
+| `<.*>`     | `<[^>]*>`      |
+| `\((.*)\)` | `\([^)]*\)`    |
+| `price=.*` | `price=[0-9]+` |
+
+---
+
+### 2. Anchor When You Require Full Matches
+
+FSMs only do what’s specified. Explicitly anchor when matching entire strings:
```

Comment:
this isn't about doing only what's specified, it's about writing what we mean. without the anchors, you'd allow a prefix and a suffix. if you mean that, great. if you don't mean that, we're advising to not write them